src/charm.py

```diff
@@ -1,103 +1,202 @@
+        self._update(event)
 
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
+    def _validate(self):
+        """Validate that configuration and relations are valid and ready.
 
-        Learn more about config at https://juju.is/docs/sdk/config
+        Raises:
+            ValueError: in case of invalid configuration.
         """
-        # Fetch the new config value
         log_level = self.model.config["log-level"].lower()
+        if log_level not in VALID_LOG_LEVELS:
+            raise ValueError(f"config: invalid log level {log_level!r}")
+
+        self._check_required_config(REQUIRED_CHARM_CONFIG)
+
+        if self.config["auth-enabled"]:
+            if not self.config["auth-provider"]:
+                raise ValueError("Invalid config: auth-provider value missing")
+
+            if not self.config["auth-provider"] in SUPPORTED_AUTH_PROVIDERS:
+                raise ValueError("Invalid config: auth-provider not supported")
+
+            if self.config["auth-provider"] == "candid":
+                self._check_required_config(REQUIRED_CANDID_CONFIG)
 
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+            if self.config["auth-provider"] == "google":
+                self._check_required_config(REQUIRED_OIDC_CONFIG)
+
+    def _check_required_config(self, config_list):
+        """Check if required config has been set by user.
+
+        Args:
+            config_list: list of required config parameters.
+
+        Raises:
+            ValueError: if any of the required config is not set.
+        """
+        for param in config_list:
+            if self.config[param].strip() == "":
```

Comment:
Is there a possibility that the key is missing from the received configuration? I am not too well-versed in charms. If so, would it be a good idea to catch the exception to re-emit it better formatted?

Reply:
In juju, you cannot remove a config, you can only unset it (sending it back to its default value). So there will be no need to check if the key is missing from the configuration.

Reply:
Understood.